File: easyplot/scripts/easyplot_palettes.py
"""Unified EasyPlot palettes from the frozen ../assets/palettes.json registry.

IDs and filters are exact and case-sensitive, including percent-encoded IDs.
The three base getters use only the standard library and never access a network.
The registry is a process-local snapshot, read on the first getter call.

Selection rules:
* n=None returns all stored colours, for every selection policy.
* prefix takes the first n colours and refuses requests beyond capacity.
* native uses the stored scheme of exactly n colours. For n=1 or 2, take the
  first n colours of the smallest stored scheme. Missing sizes are errors.
* lut samples stored entries, without interpolation: n=1 uses zero-based
  floor((length-1)/2 + 0.5); n>=2 uses floor(i*(length-1)/(n-1) + 0.5).
  Both endpoints are included. Up to 65536 samples are allowed; requesting more
  samples than stored entries repeats colours and adds no colour precision.
reverse reverses the selected subset in all cases.

Use explicit group-to-colour mappings for qualitative data. easyplot_cmap lazily
imports Matplotlib and sets missing values to #E2E2E2. Declare normalization and
limits explicitly for continuous data, including a meaningful diverging centre.
"""
from __future__ import annotations

from copy import deepcopy
from functools import lru_cache
import json
from math import floor
from numbers import Integral
from pathlib import Path
import re

__all__ = ["easyplot_palettes", "easyplot_palette_info", "easyplot_palette", "easyplot_cmap"]
REGISTRY_PATH = Path(__file__).resolve().parents[1] / "assets" / "palettes.json"


def _text(value):
    return isinstance(value, str) and bool(value.strip())


def _check_colours(colours, context):
    if not isinstance(colours, list) or not colours or any(
        not isinstance(colour, str) or re.fullmatch(r"#[0-9A-Fa-f]{6}", colour) is None
        for colour in colours
    ):
        raise ValueError(f"Invalid #RRGGBB colours in {context}")


@lru_cache(maxsize=1)
def _registry():
    data = json.loads(REGISTRY_PATH.read_text(encoding="utf-8"))
    if not isinstance(data, dict) or type(data.get("schema_version")) is not int or data["schema_version"] != 1:
        raise ValueError("Unsupported palette registry schema_version; expected 1")
    if not isinstance(data.get("palettes"), list):
        raise ValueError("Palette registry palettes must be an array")
    records = {}
    for record in data["palettes"]:
        if not isinstance(record, dict) or any(not _text(record.get(key)) for key in ("id", "label", "family")):
            raise ValueError("Each palette needs a nonempty id, label and family")
        id = record["id"]
        if id in records:
            raise ValueError(f"Duplicate palette id: {id!r}")
        if record.get("kind") not in ("qualitative", "sequential", "diverging", "cyclic", "multisequential"):
            raise ValueError(f"Invalid palette kind for {id!r}")
        if record.get("selection") not in ("prefix", "native", "lut"):
            raise ValueError(f"Invalid palette selection for {id!r}")
        _check_colours(record.get("colours"), id)
        cvd = record.get("cvd")
        if not isinstance(cvd, dict) or cvd.get("status") not in ("reported", "conditional", "not_assessed", "not_recommended"):
            raise ValueError(f"Invalid cvd status for {id!r}")
        if any(not isinstance(cvd.get(key), str) for key in ("evidence", "note")):
            raise ValueError(f"Missing cvd evidence or note for {id!r}")
        if record["selection"] == "native":
            schemes = record.get("native_sizes")
            if not isinstance(schemes, dict) or not schemes:
                raise ValueError(f"Missing stored native sizes for {id!r}")
            for size, colours in schemes.items():
                _check_colours(colours, f"{id} native size {size}")
                if re.fullmatch(r"[1-9][0-9]*", size) is None or len(colours) != int(size):
                    raise ValueError(f"Invalid stored native size {size!r} for {id!r}")
        records[id] = record
    return records


def _record(id):
    if not _text(id) or id not in _registry():
        raise ValueError(f"Unknown palette id: {id!r}; use easyplot_palettes() for exact IDs")
    return _registry()[id]


def _capacity(record):
    if record["selection"] == "lut":
        return 65536
    if record["selection"] == "native":
        return max(map(int, record["native_sizes"]))
    return len(record["colours"])


@lru_cache(maxsize=1)
def _metadata():
    return [dict(
        **{key: record[key] for key in ("id", "label", "family", "kind", "selection")},
        n_colours=len(record["colours"]), max_n=_capacity(record),
        cvd_status=record["cvd"]["status"], cvd_evidence=record["cvd"]["evidence"],
        cvd_note=record["cvd"]["note"],
    ) for record in _registry().values()]


def easyplot_palettes(family=None, kind=None, cvd=None):
    """Return metadata dictionaries in registry order, with exact scalar filters.

    cvd filters a status string, not a boolean safety claim. Unknown strings match
    no records. Columns match the R data.frame: id, label, family, kind, selection,
    n_colours, max_n, cvd_status, cvd_evidence, cvd_note. max_n is the largest
    explicit request; native sizes can have gaps. Full provenance, tags and
    optional fields are available through easyplot_palette_info().
    """
    for key, value in (("family", family), ("kind", kind), ("cvd", cvd)):
        if value is not None and not _text(value):
            raise ValueError(f"{key} must be a nonempty string or None")
    return [dict(row) for row in _metadata()
            if (family is None or row["family"] == family)
            and (kind is None or row["kind"] == kind)
            and (cvd is None or row["cvd_status"] == cvd)]


def easyplot_palette_info(id):
    """Return a full defensive copy of a record, including provenance and notes."""
    return deepcopy(_record(id))


def easyplot_palette(id, n=None, reverse=False):
    """Return fresh colours using the selection rules in this module's docstring.

    n must be a positive Integral (bool and floats are rejected); reverse must
    be a Python bool. No aliases, ID decoding, recycling or interpolation occurs.
    LUT resampling may repeat stored colours, with no added precision.
    """
    record = _record(id)
    if type(reverse) is not bool:
        raise ValueError("reverse must be a boolean")
    if n is not None and (isinstance(n, bool) or not isinstance(n, Integral) or n < 1):
        raise ValueError("n must be a positive integer")
    colours = record["colours"]
    if n is None:
        result = list(colours)
    else:
        n = int(n)
        capacity = _capacity(record)
        if n > capacity:
            raise ValueError(f"{id} supports at most {capacity} requested colours")
        if record["selection"] == "prefix":
            result = colours[:n]
        elif record["selection"] == "native":
            schemes = record["native_sizes"]
            size = str(min(map(int, schemes))) if n < 3 else str(n)
            if size not in schemes or len(schemes[size]) < n:
                raise ValueError(f"No stored native size {n} for {id}; available sizes: {', '.join(schemes)}")
            result = schemes[size][:n]
        else:
            indices = [len(colours) // 2] if n == 1 else [
                floor(i * (len(colours) - 1) / (n - 1) + 0.5) for i in range(n)
            ]
            result = [colours[index] for index in indices]
    return result[::-1] if reverse else result


def easyplot_cmap(id, reverse=False):
    """Return a lazy Matplotlib ListedColormap; qualitative palettes are refused."""
    if _record(id)["kind"] == "qualitative":
        raise ValueError("A continuous cmap cannot use a qualitative palette")
    colours = easyplot_palette(id, reverse=reverse)
    from matplotlib.colors import ListedColormap
    cmap = ListedColormap(colours, name=f"easyplot_{id}" + ("_r" if reverse else ""))
    cmap.set_bad("#E2E2E2")
    return cmap

File: easyplot/scripts/easyplot_palettes.R
# Unified EasyPlot palettes. Source this file to load ../assets/palettes.json
# using installed jsonlite. All IDs and filters are exact, case-sensitive strings,
# including percent-encoded IDs. There is no network or external palette package.
# The registry is a snapshot loaded once per source(); R copy-on-modify semantics
# and a locked private environment isolate returned records and colour vectors.
#
# easyplot_palettes(family = NULL, kind = NULL, cvd = NULL) returns a data.frame
# in registry order with id, label, family, kind, selection, n_colours, max_n,
# cvd_status, cvd_evidence, cvd_note. Filters are nonempty scalar strings; unknown
# strings return zero rows. cvd is a status, not a logical safety claim. max_n is
# the largest explicit request; native sizes can have gaps.
# easyplot_palette_info(id) returns every record field, including tags and source.
#
# easyplot_palette(id, n = NULL, reverse = FALSE):
# * NULL returns every stored colour, for all selection policies.
# * prefix takes the first n colours, refusing requests beyond capacity.
# * native takes the exact stored n-colour scheme. n=1 or 2 takes the first n
#   colours of the smallest stored scheme. Unstored sizes raise an error.
# * lut selects stored entries without interpolation. The zero-based index for
#   n=1 is floor((length-1)/2 + 0.5); for n>=2 it is
#   floor(i*(length-1)/(n-1) + 0.5), including both endpoints. Up to 65536
#   samples are allowed. Repeated colours add no precision beyond the table.
# reverse reverses the selected subset. n accepts a finite, positive, integer-
# valued numeric scalar (including ordinary R literals such as 3); reverse must
# be a single nonmissing logical. Logical n, vectors and classed n are rejected.
#
# scale_fill_easyplot(id, levels, ..., reverse = FALSE) and scale_colour_easyplot
# accept only qualitative palettes and explicit unique, nonempty character levels.
# Named values plus fixed limits preserve group identity, with drop=FALSE by
# default and missing values #E2E2E2. ggplot2 is required only when these are used.
# For continuous data, use ggplot2 directly and declare limits/normalization:
# ggplot2::scale_fill_gradientn(colours = easyplot_palette("viridis.viridis"),
#                              limits = c(0, 1), na.value = "#E2E2E2")
# gradientn interpolates while rendering; use a stepped/binned scale when original
# source swatches must remain exact. Declare a meaningful centre for diverging data.

local({
  if (!requireNamespace("jsonlite", quietly = TRUE)) {
    stop("EasyPlot palettes require the installed jsonlite package.", call. = FALSE)
  }
  source_frames <- Filter(function(frame) !is.null(frame$ofile), sys.frames())
  if (!length(source_frames)) stop("Load easyplot_palettes.R with source().", call. = FALSE)
  source_frame <- source_frames[[length(source_frames)]]
  script_path <- if (isTRUE(source_frame$chdir)) basename(source_frame$ofile) else source_frame$ofile
  script_path <- normalizePath(script_path, mustWork = TRUE)
  registry <- jsonlite::fromJSON(
    file.path(dirname(script_path), "..", "assets", "palettes.json"), simplifyVector = FALSE
  )
  text_scalar <- function(value) {
    is.character(value) && length(value) == 1L && !is.na(value) &&
      is.null(dim(value)) && !is.object(value) && nzchar(trimws(value))
  }
  check_colours <- function(values, context) {
    if (!is.list(values) || !length(values) || !all(vapply(values, function(value) {
      text_scalar(value) && grepl("^#[0-9A-Fa-f]{6}$", value)
    }, logical(1L)))) stop("Invalid #RRGGBB colours in ", context, call. = FALSE)
  }
  if (!is.list(registry) || !identical(registry$schema_version, 1L)) {
    stop("Unsupported palette registry schema_version; expected 1.", call. = FALSE)
  }
  if (!is.list(registry$palettes)) stop("Palette registry palettes must be an array.", call. = FALSE)
  for (record in registry$palettes) {
    if (!is.list(record) || !all(vapply(record[c("id", "label", "family")], text_scalar, logical(1L)))) {
      stop("Each palette needs a nonempty id, label and family.", call. = FALSE)
    }
    if (!text_scalar(record$kind) || !record$kind %in% c("qualitative", "sequential", "diverging", "cyclic", "multisequential")) {
      stop("Invalid palette kind for ", record$id, call. = FALSE)
    }
    if (!text_scalar(record$selection) || !record$selection %in% c("prefix", "native", "lut")) {
      stop("Invalid palette selection for ", record$id, call. = FALSE)
    }
    check_colours(record$colours, record$id)
    if (!is.list(record$cvd) || !text_scalar(record$cvd$status) ||
        !record$cvd$status %in% c("reported", "conditional", "not_assessed", "not_recommended")) {
      stop("Invalid cvd status for ", record$id, call. = FALSE)
    }
    if (!all(vapply(record$cvd[c("evidence", "note")], function(value) {
      is.character(value) && length(value) == 1L && !is.na(value)
    }, logical(1L)))) stop("Missing cvd evidence or note for ", record$id, call. = FALSE)
    if (record$selection == "native") {
      schemes <- record$native_sizes
      if (!is.list(schemes) || !length(schemes) || is.null(names(schemes)) ||
          anyDuplicated(names(schemes)) || any(!grepl("^[1-9][0-9]*$", names(schemes)))) {
        stop("Missing or invalid stored native sizes for ", record$id, call. = FALSE)
      }
      for (size in names(schemes)) {
        check_colours(schemes[[size]], paste(record$id, "native size", size))
        if (length(schemes[[size]]) != as.double(size)) {
          stop("Invalid stored native size ", size, " for ", record$id, call. = FALSE)
        }
      }
    }
  }
  ids <- vapply(registry$palettes, function(record) record$id, character(1L))
  if (anyDuplicated(ids)) stop("Duplicate palette id in registry.", call. = FALSE)
  palettes <- stats::setNames(registry$palettes, ids)
  capacity <- function(record) {
    if (record$selection == "lut") return(65536L)
    if (record$selection == "native") return(max(as.integer(names(record$native_sizes))))
    length(record$colours)
  }
  get_record <- function(id) {
    if (!text_scalar(id) || !id %in% ids) {
      stop("Unknown palette id; use easyplot_palettes() for exact IDs.", call. = FALSE)
    }
    palettes[[id, exact = TRUE]]
  }
  metadata <- as.data.frame(stats::setNames(lapply(c("id", "label", "family", "kind", "selection"), function(key) {
    vapply(palettes, function(record) record[[key]], character(1L), USE.NAMES = FALSE)
  }), c("id", "label", "family", "kind", "selection")), stringsAsFactors = FALSE)
  metadata$n_colours <- vapply(palettes, function(record) length(record$colours), integer(1L), USE.NAMES = FALSE)
  metadata$max_n <- vapply(palettes, capacity, integer(1L), USE.NAMES = FALSE)
  for (key in c("status", "evidence", "note")) {
    metadata[[paste0("cvd_", key)]] <- vapply(palettes, function(record) record$cvd[[key]], character(1L), USE.NAMES = FALSE)
  }

  easyplot_palettes <- function(family = NULL, kind = NULL, cvd = NULL) {
    filters <- list(family = family, kind = kind, cvd = cvd)
    for (key in names(filters)) {
      if (!is.null(filters[[key]]) && !text_scalar(filters[[key]])) {
        stop(key, " must be a nonempty string or NULL.", call. = FALSE)
      }
    }
    keep <- rep(TRUE, nrow(metadata))
    if (!is.null(family)) keep <- keep & metadata$family == family
    if (!is.null(kind)) keep <- keep & metadata$kind == kind
    if (!is.null(cvd)) keep <- keep & metadata$cvd_status == cvd
    result <- metadata[keep, , drop = FALSE]
    rownames(result) <- NULL
    result
  }

  easyplot_palette_info <- function(id) get_record(id)

  easyplot_palette <- function(id, n = NULL, reverse = FALSE) {
    record <- get_record(id)
    if (!is.logical(reverse) || length(reverse) != 1L || is.na(reverse) ||
        !is.null(dim(reverse)) || is.object(reverse)) {
      stop("reverse must be TRUE or FALSE.", call. = FALSE)
    }
    if (!is.null(n) && (!typeof(n) %in% c("integer", "double") || length(n) != 1L ||
        is.object(n) || !is.null(dim(n)) || !is.finite(n) || n < 1 || n != floor(n))) {
      stop("n must be one positive integer.", call. = FALSE)
    }
    values <- unlist(record$colours, use.names = FALSE)
    if (!is.null(n)) {
      maximum <- capacity(record)
      if (n > maximum) stop(id, " supports at most ", maximum, " requested colours.", call. = FALSE)
      if (record$selection == "prefix") {
        values <- values[seq_len(n)]
      } else if (record$selection == "native") {
        size <- if (n < 3) min(as.integer(names(record$native_sizes))) else n
        scheme <- record$native_sizes[[as.character(size), exact = TRUE]]
        if (is.null(scheme) || length(scheme) < n) {
          stop("No stored native size ", n, " for ", id, "; available sizes: ",
               paste(names(record$native_sizes), collapse = ", "), call. = FALSE)
        }
        values <- unlist(scheme, use.names = FALSE)[seq_len(n)]
      } else {
        indices <- if (n == 1) floor(length(values) / 2) else
          floor((seq_len(n) - 1) * (length(values) - 1) / (n - 1) + 0.5)
        values <- values[indices + 1L]
      }
    }
    if (reverse) rev(values) else values
  }

  manual_scale <- function(aesthetic, id, levels, reverse, ...) {
    if (!is.character(levels) || !length(levels) || anyNA(levels) ||
        any(!nzchar(trimws(levels))) || anyDuplicated(levels) ||
        !is.null(dim(levels)) || is.object(levels)) {
      stop("levels must be explicit, nonempty, unique character values.", call. = FALSE)
    }
    if (get_record(id)$kind != "qualitative") {
      stop("EasyPlot manual scales require a qualitative palette.", call. = FALSE)
    }
    levels <- unname(levels)
    values <- stats::setNames(easyplot_palette(id, length(levels), reverse), levels)
    dots <- list(...)
    if (any(names(dots) %in% c("values", "limits", "aesthetics", "palette", "na.value"))) {
      stop("values, limits, aesthetics, palette and na.value are controlled by EasyPlot.", call. = FALSE)
    }
    if (!"drop" %in% names(dots)) dots$drop <- FALSE
    if (!requireNamespace("ggplot2", quietly = TRUE)) {
      stop("EasyPlot manual scales require the installed ggplot2 package.", call. = FALSE)
    }
    scale <- if (aesthetic == "fill") ggplot2::scale_fill_manual else ggplot2::scale_colour_manual
    do.call(scale, c(list(values = values, limits = levels, na.value = "#E2E2E2"), dots))
  }
  scale_fill_easyplot <- function(id, levels, ..., reverse = FALSE) {
    manual_scale("fill", id, levels, reverse, ...)
  }
  scale_colour_easyplot <- function(id, levels, ..., reverse = FALSE) {
    manual_scale("colour", id, levels, reverse, ...)
  }

  for (name in c("easyplot_palettes", "easyplot_palette_info", "easyplot_palette",
                 "scale_fill_easyplot", "scale_colour_easyplot")) {
    assign(name, get(name), envir = parent.env(environment()))
  }
  lockEnvironment(environment(), bindings = TRUE)
})
